casual/casual-jca/src/test/groovy/se/laz/casual/jca/pool/ReferenceCountedNetworkConnectionTest.groovy

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2022, The casual project. All rights reserved.
+ *
+ * This software is licensed under the MIT license, https://opensource.org/licenses/MIT
+ */
+package se.laz.casual.jca.pool
+
+import se.laz.casual.network.outbound.NettyNetworkConnection
+import spock.lang.Specification
+
+class ReferenceCountedNetworkConnectionTest extends Specification
+{
+   def 'test only 1 reference count ( initial )'()
+   {
+      given:
+      def networkConnection = Mock(NettyNetworkConnection) {
+         1 * close()
+      }
+      def closeListener = Mock(ReferenceCountedNetworkCloseListener){
+         1 * closed(_)
+      }
+      // 1st reference count on creation
+      ReferenceCountedNetworkConnection instance = ReferenceCountedNetworkConnection.of(networkConnection, closeListener)
+      when:
+      instance.close()
+      then:
+      noExceptionThrown()
+   }
+
+   def 'test reference counting and effects when the last reference is closed - more than initial reference'()
+   {
+      given:
+      def subsequentReferences = 4
+      def expectedNumberOfReferences = 5
+      def networkConnection = Mock(NettyNetworkConnection) {
+         1 * close()
+      }
+      def closeListener = Mock(ReferenceCountedNetworkCloseListener){
+         1 * closed(_)
+      }
+      // 1st reference count on creation
+      ReferenceCountedNetworkConnection instance = ReferenceCountedNetworkConnection.of(networkConnection, closeListener)
```

Comment:
This test sequential increments and decrements. Should we add a test using an Executor with multiple threads to ensure it's fine too?

Reply:
Yeah, that won't hurt